lib/libutils/ext/ubsan.c

```diff
@@ -19,12 +20,30 @@
 # define _ubsan_panic() _utee_panic(TEE_ERROR_GENERIC)
 #endif
 
+#define UBSAN_LOC_REPORTED (UINT32_C(1) << 31)
+
 struct source_location {
 	const char *file_name;
 	uint32_t line;
-	uint32_t column;
+	union {
+		uint32_t column;
+		uint32_t reported;
+	};
 };
 
+static bool was_already_reported(struct source_location *loc)
+{
+	uint32_t reported;
+
+	reported = loc->reported;
+	if (reported & UBSAN_LOC_REPORTED) {
```

Comment:
Is this an optimization to avoid a potentially expensive barrier in `atomic_cas_u32()`?

Reply:
Not really. This is just a part of the logic to get correct result. We need "old value" here for CAS, where "reported" flag is cleared. So we should either check that this flag is cleared or clear it directly.